lib/internal/util/comparisons.js

```diff
@@ -374,23 +374,48 @@ function setHasEqualElement(set, val1, strict, memo) {
   return false;
 }
 
-// Note: we currently run this multiple times for each loose key!
-// This is done to prevent slowing down the average case.
-function setHasLoosePrim(a, b, val) {
-  const altValues = findLooseMatchingPrimitives(val);
-  if (altValues === undefined)
-    return false;
+// See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#Loose_equality_using
+// Sadly it is not possible to detect corresponding values properly in case the
+// type is a string, number, bigint or boolean. The reason is that those values
+// can match lots of different string values (e.g., 1n == '+00001').
+function findLooseMatchingPrimitives(prim) {
+  switch (typeof prim) {
```

Comment:
Are you sure this doesn't cause performance issues still?

Reply:
It does. Significantly for loose comparison for any keys that are primitives that are not `null`, `undefined`, symbols and strings that are not loosely equal to any other values.

For strings as primitives that are not loosely equal to numbers:
(A small performance increase)

```
 assert/deepequal-set.js method='deepEqual_mixed' strict=0 len=500 n=500                           -0.38 %       ±0.93% ±1.24% ±1.62%
 assert/deepequal-set.js method='deepEqual_mixed' strict=1 len=500 n=500                   ***      5.23 %       ±1.89% ±2.53% ±3.30%
 assert/deepequal-set.js method='deepEqual_objectOnly' strict=0 len=500 n=500                      -0.32 %       ±1.25% ±1.67% ±2.19%
 assert/deepequal-set.js method='deepEqual_objectOnly' strict=1 len=500 n=500                       0.89 %       ±2.38% ±3.19% ±4.22%
 assert/deepequal-set.js method='deepEqual_primitiveOnly' strict=0 len=500 n=500                   -0.05 %       ±1.81% ±2.41% ±3.14%
 assert/deepequal-set.js method='deepEqual_primitiveOnly' strict=1 len=500 n=500                   -0.04 %       ±2.02% ±2.70% ±3.53%
 assert/deepequal-set.js method='notDeepEqual_mixed' strict=0 len=500 n=500                ***      2.59 %       ±0.99% ±1.32% ±1.72%
 assert/deepequal-set.js method='notDeepEqual_mixed' strict=1 len=500 n=500                         2.00 %       ±2.38% ±3.17% ±4.12%
 assert/deepequal-set.js method='notDeepEqual_objectOnly' strict=0 len=500 n=500                   -0.25 %       ±0.84% ±1.12% ±1.46%
 assert/deepequal-set.js method='notDeepEqual_objectOnly' strict=1 len=500 n=500                   -0.34 %       ±2.01% ±2.68% ±3.49%
 assert/deepequal-set.js method='notDeepEqual_primitiveOnly' strict=0 len=500 n=500        ***      4.24 %       ±1.89% ±2.52% ±3.28%
 assert/deepequal-set.js method='notDeepEqual_primitiveOnly' strict=1 len=500 n=500                 0.57 %       ±3.65% ±4.87% ±6.37%
```

For numbers as primitives:
(A significant performance loose for loose not equal checks)

```
 assert/deepequal-set.js method='deepEqual_mixed' strict=0 len=500 n=500                     *     -4.01 %       ±3.39% ±4.60% ±6.18%
 assert/deepequal-set.js method='deepEqual_mixed' strict=1 len=500 n=500                   ***      4.35 %       ±1.94% ±2.61% ±3.44%
 assert/deepequal-set.js method='deepEqual_primitiveOnly' strict=0 len=500 n=500             *      5.06 %       ±4.38% ±5.93% ±7.91%
 assert/deepequal-set.js method='deepEqual_primitiveOnly' strict=1 len=500 n=500                    0.48 %       ±5.28% ±7.08% ±9.31%
 assert/deepequal-set.js method='notDeepEqual_mixed' strict=0 len=500 n=500                ***    -87.74 %       ±3.30% ±4.49% ±6.05%
 assert/deepequal-set.js method='notDeepEqual_mixed' strict=1 len=500 n=500                        -0.52 %       ±2.56% ±3.44% ±4.55%
 assert/deepequal-set.js method='notDeepEqual_primitiveOnly' strict=0 len=500 n=500        ***    -39.88 %       ±4.70% ±6.31% ±8.34%
 assert/deepequal-set.js method='notDeepEqual_primitiveOnly' strict=1 len=500 n=500                -2.68 %       ±3.50% ±4.72% ±6.28%
```

Reply:
I tried another approach to overcome the downside but it is simply not possible to absolutely be sure there is no other loosely equal entry.

Now a primitive that could match something else has to go throw all entries at least once. Before, it would stop when the entry was found as not having a corresponding entry.

Reply:
What I was referring to was specifically the use of `switch (typeof prim)` vs. an `if`-`else` ladder. I'm thinking V8 might still not optimize well when `typeof` is used in this way, because it's being treated as a variable instead of a direct comparison?

Reply:
Tracking issue: https://bugs.chromium.org/p/v8/issues/detail?id=8093

Reply:
Seems like there is a tiny difference. I don't think it's significant enough that we should refactor the code. Instead, V8 should just improve it and we'll benefit from it as soon as that lands in Node.